### PR TITLE
Support ROS2 build

### PR DESCRIPTION
**Pull request by kenji-miyake**

### CODE REVIEW

**Comment by kenji-miyake** [comment]
It seems action-ros-ci doesn't install dependencies that have `condition` of `ROS_VERSION `? :thinking:

**Comment by kenji-miyake** [comment]
Seeing the code of action-ros-ci, it's a bit annoying to fix this problem... :pleading_face: (because of ROS1 and Windows support)

**Comment by kenji-miyake** [comment]
Posted a PR. https://github.com/ros-tooling/action-ros-ci/pull/697

File: docker/melodic/Dockerfile [hunk 1]
```diff
@@ -11,8 +10,8 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
 RUN mkdir -p /root/catkin_ws/src
 WORKDIR /root/catkin_ws/src
 RUN /bin/bash -c '. /opt/ros/melodic/setup.bash; catkin_init_workspace'
-RUN git clone https://github.com/SMRT-AIST/fast_gicp.git --recursive
-RUN git clone https://github.com/koide3/hdl_graph_slam.git
+RUN git clone https://github.com/SMRT-AIST/fast_gicp.git --recursive --depth=1
```
Comment (KeisukeShima) [inline docker/melodic/Dockerfile:14]: I added the `--depth=1` option to avoid `git clone` errors.

https://github.com/KeisukeShima/ndt_omp/runs/4639280919?check_suite_focus=true#step:4:3179

**Review by KeisukeShima — APPROVED** [review]
LGTM

**Comment by KeisukeShima** [comment]
@kenji-miyake It looks good to me. Please check my changes before you merge this PR.

File: CMakeLists.txt [hunk 2]
```diff
@@ -25,19 +98,19 @@ find_package(OpenMP)
...
   src/pclomp/ndt_omp.cpp
   src/pclomp/gicp_omp.cpp
 )
 
-target_link_libraries(ndt_omp PUBLIC ${PCL_LIBRARIES})
+target_link_libraries(ndt_omp ${PCL_LIBRARIES})
 
 if(OpenMP_CXX_FOUND)
-  target_link_libraries(ndt_omp PUBLIC OpenMP::OpenMP_CXX)
+  target_link_libraries(ndt_omp OpenMP::OpenMP_CXX)
```
Comment (kenji-miyake) [inline CMakeLists.txt:111]: Why did you remove `PUBLIC`?

File: CMakeLists.txt [hunk 3]
```diff
@@ -25,19 +98,19 @@ find_package(OpenMP)
 ###########
 ## Build ##
 ###########
-
```
Comment (kenji-miyake) [inline CMakeLists.txt]: Please revert this line to keep the diff minimum.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -32,3 +32,38 @@ jobs:
           file: ${{github.workspace}}/docker/${{ matrix.ROS_DISTRO }}/Dockerfile
           context: .
           push: false
+
+  build-ros2:
```
Comment (kenji-miyake) [inline .github/workflows/build.yml]: Why did you add this job? What's the difference between `build` job?

Comment (kenji-miyake) [reply .github/workflows/build.yml]: Ah, this was added by me? So can you remove this?

File: CMakeLists.txt (hunk 2 above)
Comment (KeisukeShima) [reply CMakeLists.txt:111]: This is because the docker CI for ROS2 failed with the following error.
And the solution to this error is to remove `PUBLIC`.
<https://stackoverflow.com/questions/59522267/cmake-rejects-a-second-target-link-libraries-talking-about-keyword-vs-plain>

```
Starting >>> ndt_omp
--- stderr: ndt_omp
CMake Warning (dev) at CMakeLists.txt:87 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  CMake variable PCL_ROOT is set to:

    /usr

  For compatibility, CMake is ignoring the variable.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at CMakeLists.txt:107 (target_link_libraries):
  The plain signature for target_link_libraries has already been used with
  the target "ndt_omp".  All uses of target_link_libraries with a target must
  be either all-keyword or all-plain.

  The uses of the plain signature are here:

   * /opt/ros/foxy/share/ament_cmake_target_dependencies/cmake/ament_target_dependencies.cmake:145 (target_link_libraries)



---
Failed   <<< ndt_omp [9.05s, exited with code 1]
```

File: CMakeLists.txt (hunk 3 above)
Comment (KeisukeShima) [reply CMakeLists.txt]: Added new line 36a9ad5

File: .github/workflows/build.yml (hunk 4 above)
Comment (KeisukeShima) [reply .github/workflows/build.yml]: > Ah, this was added by me? So can you remove this?

Yes, you added this job. So I left it in place.
OK, I will remove this job.

Comment (KeisukeShima) [reply .github/workflows/build.yml]: de1c6ed

File: CMakeLists.txt (hunk 2 above)
Comment (kenji-miyake) [reply CMakeLists.txt:111]: Is this because of `ament_auto_add_library`?
I feel `all-keyword` is better if possible.

File: .github/workflows/build.yml (hunk 4 above)
Comment (kenji-miyake) [reply .github/workflows/build.yml]: Thanks!

File: CMakeLists.txt (hunk 2 above)
Comment (kenji-miyake) [reply CMakeLists.txt:111]: Will this work?
https://github.com/ament/ament_cmake/blob/4ac1ed1153ac774116c273021944eb3ea836495c/ament_cmake_auto/cmake/ament_auto_add_library.cmake#L66

```cmake
ament_auto_add_library(ndt_omp SHARED PUBLIC
  src/pclomp/voxel_grid_covariance_omp.cpp
  src/pclomp/ndt_omp.cpp
  src/pclomp/gicp_omp.cpp
)

target_link_libraries(ndt_omp PUBLIC ${PCL_LIBRARIES})

if(OpenMP_CXX_FOUND)
  target_link_libraries(ndt_omp PUBLIC OpenMP::OpenMP_CXX)
```

Comment (kenji-miyake) [reply CMakeLists.txt:111]: Ah, but since the upstream's style is `plain`, I think it's okay.
https://github.com/koide3/ndt_omp/blob/399d61318a471758234e0d747ab4e407c74fcb76/CMakeLists.txt

**Comment by kenji-miyake** [comment]
@wep21 Could you please double-check? :pray:

File: CMakeLists.txt (hunk 2 above)
Comment (KeisukeShima) [reply CMakeLists.txt:111]: > Will this work? https://github.com/ament/ament_cmake/blob/4ac1ed1153ac774116c273021944eb3ea836495c/ament_cmake_auto/cmake/ament_auto_add_library.cmake#L66
> 
> ```cmake
> ament_auto_add_library(ndt_omp SHARED PUBLIC
>   src/pclomp/voxel_grid_covariance_omp.cpp
>   src/pclomp/ndt_omp.cpp
>   src/pclomp/gicp_omp.cpp
> )
> 
> target_link_libraries(ndt_omp PUBLIC ${PCL_LIBRARIES})
> 
> if(OpenMP_CXX_FOUND)
>   target_link_libraries(ndt_omp PUBLIC OpenMP::OpenMP_CXX)
> ```

It doesn't work. :cry: 

> Ah, but since the upstream's style is plain, I think it's okay.

Thanks for letting me know. I'll keep the current code.

Comment (kenji-miyake) [reply CMakeLists.txt:111]: :pleading_face: 
Okay, thank you!

**Review by wep21 — APPROVED** [review]
looks good to me

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -14,7 +14,7 @@ jobs:
     runs-on: ubuntu-latest
```
Comment (kenji-miyake) [inline .github/workflows/build.yml:13]: It seems no check is running for this PR.
I guess you should fix the branch names in `on` conditions?

Comment (kenji-miyake) [reply .github/workflows/build.yml:13]: Or removing branches from `pull_request` might be good.

Comment (KeisukeShima) [reply .github/workflows/build.yml:13]: Now the CI works on this PR! 0246885
